internal/training/grpo_engine/policy.c

```
@@ -0,0 +1,874 @@
+/* ─── Helper: Dequantize a single row from embedding table ─── */
+static void dequant_embed_row(float *out, const void *embed_table, int embed_dtype,
+                               int token, int hidden_dim) {
+    if (embed_dtype == 0) {  /* F32 */
+        const float *embed_f32 = (const float *)embed_table;
+        memcpy(out, embed_f32 + (size_t)token * (size_t)hidden_dim,
+               (size_t)hidden_dim * sizeof(float));
+    } else {
+        /* Quantized: treat as 1-row matmul with identity vector */
+        /* Create identity input: [1.0, 1.0, ...] */
+        float *identity = (float *)malloc((size_t)hidden_dim * sizeof(float));
+        for (int i = 0; i < hidden_dim; i++) identity[i] = 1.0f;
```

Cppcheck (2.21.0)

[warning] If memory allocation fails, then there is a possible null pointer dereference

(nullPointerOutOfMemory)